### PR TITLE
Copy all server test assets to tmp before running tests

### PR DESCRIPTION
**Pull request by elpaso**
this should fix https://github.com/qgis/QGIS/pull/48904#issuecomment-1152887478

### CODE REVIEW

**Comment by elpaso** [comment]
CC @strk

**Comment by strk** [comment]
I'm going to merge your commit into the PR having CI check for files modifications. You ok with me rewording the commit log to close #48978 right ?

File: tests/src/python/test_qgsserver.py [hunk 1]
```diff
@@ -117,9 +118,17 @@ def setUp(self):
         self.fontFamily = QgsFontUtils.standardTestFontFamily()
         QgsFontUtils.loadStandardTestFonts(['All'])
 
-        self.testdata_path = unitTestDataPath('qgis_server') + '/'
+        self.temporary_dir = tempfile.TemporaryDirectory()
+        self.temporary_path = self.temporary_dir.name
+
+        # Copy all testdata to the temporary directory
+        copytree(unitTestDataPath('qgis_server'), os.path.join(self.temporary_path, 'qgis_server'))
+        copytree(unitTestDataPath('qgis_server_accesscontrol'), os.path.join(self.temporary_path, 'qgis_server_accesscontrol'))
```
Comment (strk) [inline tests/src/python/test_qgsserver.py:126]: Does the test really need to write to these files ?
I'm happy to see commits allowing us to enable CI checking for no source-tree modification, but I also wonder if it's overkill to make a full tree copy (for each test) if the usage pattern is read-only...

**Comment by strk** [comment]
For the record: your commit DID fix the case reported in #48978 for me:
```
[strk@c19:/usr/local/src/qgis/qgis/src/master/build(do-not-override-source-files)] xvfb-run ctest --output-on-failure -R PyQgsServerWMSGetFeatureInfoPG
Test project /usr/local/src/qgis/qgis/src/master/build
    Start   1: logGitStatus
1/3 Test   #1: logGitStatus .....................   Passed    0.01 sec
    Start 806: PyQgsServerWMSGetFeatureInfoPG
2/3 Test #806: PyQgsServerWMSGetFeatureInfoPG ...   Passed    2.87 sec
    Start   2: checkGitStatus
3/3 Test   #2: checkGitStatus ...................   Passed    0.01 sec

100% tests passed, 0 tests failed out of 3

Label Time Summary:
POSTGRES    =   2.87 sec*proc (1 test)

Total Test time (real) =   2.91 sec
```

**Comment by strk** [comment]
The failure is:
```
 ERROR: testGetFeatureInfoThousandSeparator (__main__.TestQgsServerWMSLocaleOverride)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/QGIS/tests/src/python/test_qgsserver.py", line 160, in tearDown
    self.temporary_dir.cleanup()
AttributeError: 'TestQgsServerWMSLocaleOverride' object has no attribute 'temporary_dir'
```

**Comment by strk** [comment]
@elpaso I think the fix for the CI failure is https://github.com/qgis/QGIS/pull/48904/commits/a315156fd78acb17ee152ce7f8b979ecff1dadf4 which I pushed to #48904 - let me know if you prefer having that commit in a separate PR

**Comment by strk** [comment]
If my commit fixes the problem feel free to cherry-pick it into this PR to keep the two tasks separated (my PR is about having CI guarding after those modifications, this PR is about fixing the actual modification...)

**Comment by strk** [comment]
I did push my commit here directly too

**Comment by strk** [comment]
There seem to be more subclasses NOT using the base class setup, so this PR is going to need more work. Some tests are already doing the files copies, btw, so doing a copy upfront in base class would make 2 copies for those specific tests.  I'm thinking being specific in each subtest might be better because also lighter... Regarding "You never know"... the CI PR can tell us :)

**Comment by strk** [comment]
Just a note. I belive in an old PR of mine (killed by Stale Bot) I had a commit which performed the copy upfront from cmake (using a FIXTURE I think). That way each test could still modify the files in compatible ways (stats/wals) but the upfront copy would be done once (less expensive).

Maybe we can dig that commit out of it:
https://github.com/qgis/QGIS/pull/31980/commits

The commit seems to be still there, but it wasn't using a FIXTURE (a FIXTURE would sound better to me):
https://github.com/qgis/QGIS/pull/31980/commits/95c65a74ded4c302d9c2ac8de32104a3dd4e58fa

There were also propedeutic commits used to abstract away the test data dir:
https://github.com/qgis/QGIS/pull/31980/commits/16a67feb6b268442c916d71a804b3060d6341082

**Comment by strk** [comment]
@elpaso I merged master branch into this branch and pushed another commit to fix further failures (subclasses expecting self.testdata_path ending with a slash...) - I've also changed such expectance in a subclass but my goal here is enabling CI guarding after sourcecode modfication so would not spend too much time on finding all misues of paths joining for now (although it's good to do so eventually)

**Comment by strk** [comment]
@elpaso let me know if it's ok with you if I reorganize the commits in this PR (rebase to master, reorder to fix the usage problems first, put your commit last and reword to mention fixing the right issue which is #48978)

**Comment by strk** [comment]
Next issue: TestQgsServerCacheManager is starting its own server, once for all tests, in setUpClass, but it's not calling the base class setUp method, so at the time of tearDown the temporary dir was not created.

This makes me wonder: should the server start and copy happen once for all tests instead of for each test ?

**Comment by strk** [comment]
It sounds like this snippet from project.qgs:
```
tests/testdata/qgis_server/test_opacity_project.qgs:    <layer-tree-layer checked="Qt::Checked" expanded="0" providerKey="gdal" name="landsat" source="../landsat.tif" id="landsat_a7d15b35_ca83_4b23_a9fb_af3fbdd60d15">
```

I belive you're only copying the qgis_server/ subdir, but the project in that subdir references files above it...

**Comment by strk** [comment]
I tried copying more files (one dir up) but I ended out of space... oh well

**Comment by strk** [comment]
@elpaso we're almost there. This seems to be the last leftover issue:

`The CRS is not valid`

Received by PyQgsServerWMSGetMapSizeServer test. I suspect we're missing to copy a CRS database. Ideas ?

**Comment by strk** [comment]
Nevermind, found the fix: 2c7f8674423c724295ab9b1edd24a526f0f61388

**Review by strk — APPROVED** [review]
I'm approving this, if CI is happy, I'm happy

**Comment by strk** [comment]
A squashed merge of this PR was added to PR #48904 which also enables CI checking for source tree modifications

**Pull request merged**